pkg/storage/gcs/gcp_labels_tags.go

```diff
@@ -35,3 +74,216 @@ func getUserLabels(infraLister configlisters.InfrastructureLister) (map[string]s
+	}
+	bindings := client.ListEffectiveTags(ctx, listBindingsReq)
+	// a resource can have a maximum of {gcpMaxTagsPerResource} tags attached to it.
+	// Will iterate for {gcpMaxTagsPerResource} times in the worst case scenario, if
+	// none of the break conditions are met. Should the {gcpMaxTagsPerResource} be
+	// increased in future, it should not create an issue, since this is an optimization
+	// attempt to reduce the number the tag write calls by skipping already existing tags,
+	// since it has a quota restriction.
+	for i := 0; i < gcpMaxTagsPerResource; i++ {
+		binding, err := bindings.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil || binding == nil {
+			klog.V(1).Infof("failed to list effective tags on the %s bucket: %v: %v", parent, binding, err)
+			break
+		}
+		tag := binding.GetNamespacedTagValue()
+		if _, exist := dupTags[tag]; exist {
+			dupTags[tag] = true
+			klog.V(1).Infof("filterTagList: skipping tag %s already exists on the %s bucket", tag, parent)
+		}
+	}
+
+	filteredTags := make([]string, 0, len(tagList))
+	for tagValue, dup := range dupTags {
+		if !dup {
+			filteredTags = append(filteredTags, tagValue)
+		}
+	}
+
+	return filteredTags
+}
+
+// getCreateCallOptions returns a list of additional call options to use for
+// the create operations.
+func getCreateCallOptions() []gax.CallOption {
+	return []gax.CallOption{
+		gax.WithRetry(func() gax.Retryer {
+			return gax.OnHTTPCodes(gax.Backoff{
+				Initial:    90 * time.Second,
+				Max:        5 * time.Minute,
+				Multiplier: 2,
+			},
+				http.StatusTooManyRequests)
+		}),
+	}
+}
+
+// getTagBindingsClient returns the client to be used for creating tag bindings to
+// the resources.
+func getTagBindingsClient(ctx context.Context, listers *regopclient.StorageListers, location string) (*rscmgr.TagBindingsClient, error) {
+	cfg, err := GetConfig(listers)
+	if err != nil {
+		klog.Errorf("getTagBindingsClient: failed to read gcp config: %w", err)
+		return nil, err
+	}
+
+	endpoint := fmt.Sprintf("https://%s-%s", location, resourceManagerHostSubPath)
+	opts := []option.ClientOption{
+		option.WithCredentialsJSON([]byte(cfg.KeyfileData)),
+		option.WithEndpoint(endpoint),
+	}
+
+	return rscmgr.NewTagBindingsRESTClient(ctx, opts...)
+}
+
+// addTagsToStorageBucket adds the user-defined tags in the Infrastructure resource
+// to the passed GCP bucket resource. It's wrapper around addUserTagsToStorageBucket()
+// additionally updates status condition.
+func addTagsToStorageBucket(ctx context.Context, cr *imageregistryv1.Config, listers *regopclient.StorageListers, bucketName, region string) error {
+	if err := addUserTagsToStorageBucket(ctx, listers, bucketName, region); err != nil {
+		util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionFalse,
+			gcpTagsFailedStatusReason, err.Error())
+		return err
+	}
+	util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionTrue,
+		"Tagging Bucket Successful",
```

Comment:
Either both calls should use named constants, or literal strings.

Reply:
suggestion incorporated, success reason defined as a constant.